From d65b9615e81803e98b9c3277957e360566de8bd2 Mon Sep 17 00:00:00 2001
From: WangJing <wangjing@mprc.pku.edu.cn>
Date: Mon, 10 Oct 2011 18:06:31 +0800
Subject: [PATCH 016/641] UniCore64:Add start_thread

fs/binfmt_elf.c:977:2: error: implicit declaration of function 'start_thread'

Signed-off-by: WangJing <wangjing@mprc.pku.edu.cn>
---
 arch/unicore64/include/asm/processor.h |   13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/arch/unicore64/include/asm/processor.h b/arch/unicore64/include/asm/processor.h
index 0506482..6d01653 100644
--- a/arch/unicore64/include/asm/processor.h
+++ b/arch/unicore64/include/asm/processor.h
@@ -43,4 +43,17 @@ extern unsigned long get_wchan(struct task_struct *p);
 
 #endif /* __KERNEL__ */
 
+#define start_thread(regs, pc, sp)					\
+({									\
+	unsigned long *stack = (unsigned long *)sp;			\
+	set_fs(USER_DS);						\
+	memset(regs->uc64_regs, 0, sizeof(regs->uc64_regs));		\
+	regs->UC64_ASR = USER_MODE;					\
+	regs->UC64_R31 = pc & ~1;	/* pc */                        \
+	regs->UC64_R29 = sp;		/* sp */                        \
+	regs->UC64_R02 = stack[2];	/* r2 (envp) */                 \
+	regs->UC64_R01 = stack[1];	/* r1 (argv) */                 \
+	regs->UC64_R00 = stack[0];	/* r0 (argc) */                 \
+})
+
 #endif /* __ASM_UNICORE64_PROCESSOR_H__ */
-- 
1.7.9.5

